File: problems_026to050/problem_049/solution.py
import os
import sys
from collections import Counter
from itertools import product

PROJECT_PATH = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.append(PROJECT_PATH)

from measurement import Stopwatch
from toolbox import smaller_primes


class Solution(Stopwatch):
    def __init__(self):
        super().__init__()

    def execute(self):
        answer = None
        primes_per_digits = {}

        for prime in smaller_primes(10000):
            if prime < 1000:
                continue

            digit_set = Counter(str(prime))
            combination = ''.join(sorted(digit_set.elements()))
            if combination in primes_per_digits:
                primes_per_digits[combination].append(prime)
            else:
                primes_per_digits[combination] = [prime]

        for combination in primes_per_digits:
            if len(primes_per_digits[combination]) < 3:
                continue

            permutable_primes = primes_per_digits[combination]
            for (a, b) in product(permutable_primes, repeat=2):
                if a >= b:
                    continue
                c = 2*b - a
                if c in permutable_primes and b != 4817 and a != 1487:
                    answer = str(a) + str(b) + str(c)

        return answer

if __name__ == '__main__':
    soln = Solution()
    result = soln.execute()
    print(result)

File: problems_026to050/problem_049/solution.ts
import { isPrime } from 'mathjs';

import { SolutionFunction, Dictionary } from '@/lib/types';

const solution: SolutionFunction = () => {
  let answer: ReturnType<SolutionFunction> = null;

  const primesPerDigits: Dictionary<number[]> = {};

  for (let num = 3; num < 10000; num += 2) {
    if (num >= 1000 && isPrime(num)) {
      const prime = num;
      const digitSet = String(prime).split('');
      const combination = digitSet.slice().sort().join('');

      primesPerDigits[combination] = primesPerDigits[combination] || [];
      primesPerDigits[combination].push(prime);
    }
  }

  Object.keys(primesPerDigits).every((combination) => {
    if (primesPerDigits[combination].length >= 3) {
      const permutablePrimes = primesPerDigits[combination];

      return permutablePrimes.every((a) => permutablePrimes.every((b) => {
        if (a < b) {
          const c = 2 * b - a;

          if (permutablePrimes.includes(c) && b !== 4817 && a !== 1487) {
            answer = String(a) + String(b) + String(c);
          }

          return answer === null;
        }

        return true;
      }));
    }

    return true;
  });

  return answer;
};

export default solution;
